src/main/java/org/example/email_demo/service/GithubRepoService.java

```diff
@@ -0,0 +1,155 @@
+        return githubRepoRepository.findAll();
+    }
+
+
+    public String getRepoInfo(String username, String repositoryName) throws Exception {
+
+        String apiUrl = "https://api.github.com/repos/" + username + "/" + repositoryName;
+
+        HttpClient client = HttpClient.newHttpClient();
+
+        HttpRequest request = HttpRequest.newBuilder()
+                .uri(URI.create(apiUrl))
+                .header("Accept", "application/vnd.github.v3+json")
+                .build();
+
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+
+        if (response.statusCode() == 200) {
+            return response.body();
+        } else {
+            throw new Exception("Failed to fetch repository info. HTTP Code: " + response.statusCode());
+        }
+    }
+
+    public String getRepoInfo(Long id) throws Exception {
+        Optional<GithubRepo> repo = getRepo(id);
+        if (repo.isPresent()) {
+            GithubRepo githubRepo = repo.get();
+            return getRepoInfo(githubRepo.getUsername(), githubRepo.getRepositoryName());
+        } else {
+            throw new Exception("Repository not found");
+        }
+    }
+
+    public List<RepoActivity> getRecentActivities(String username, String repositoryName) {
+        List<RepoActivity> activityList = new ArrayList<>();
+        HttpClient client = HttpClient.newHttpClient();
+
+        try {
+            String commitsUrl = "https://api.github.com/repos/" + username + "/" + repositoryName + "/commits";
+            HttpRequest commitsRequest = HttpRequest.newBuilder()
+                    .uri(URI.create(commitsUrl))
+                    .header("Accept", "application/vnd.github.v3+json")
+                    .build();
+
+            HttpResponse<String> commitsResponse = client.send(commitsRequest, HttpResponse.BodyHandlers.ofString());
+
+            if (commitsResponse.statusCode() == 200) {
+                JsonArray commitsArray = JsonParser.parseString(commitsResponse.body()).getAsJsonArray();
+                for (int i = 0; i < commitsArray.size(); i++) {
+                    var commitObj = commitsArray.get(i).getAsJsonObject();
+                    String sha = commitObj.get("sha").getAsString();
+                    String message = commitObj.getAsJsonObject("commit").get("message").getAsString();
+                    String date = commitObj.getAsJsonObject("commit").getAsJsonObject("author").get("date").getAsString();
+
+                    RepoActivity activity = new RepoActivity(sha, message, date);
+
+                    activity.setUsername(username);
+                    activity.setRepositoryName(repositoryName);
+                    activityList.add(activity);
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        try {
+            String pullsUrl = "https://api.github.com/repos/" + username + "/" + repositoryName + "/pulls?state=all";
```

Comment:
overall the making a call and processing exception is copy-paste. could be simplified